Migrate to `scikit-build-core`

cringeyburger opened the pull request:
# Description

This PR is about migrating to a new build system, `scikit-build-core` from `setuptools` and `wheel`, while incorporating necessary changes in GitHub Workflows, Benchmarks and Documentation.

CHANGES IN DISCUSSION:
- [x] Task 27: Use SUNDIALS installation script to set the environmental variables for windows --- We currently don't support Windows users to run `nox -s pybamm-requires` or `install_KLU_Sundials.py` script. Adding this feature would change a lot of things

CHANGES REQUESTED:
- [x] Task 22: Remove images from `.gitignore`
- [x] Task 23: Include the whole `sundials_KLU_libs` in `.gitignore`
- [x] Task 24: Remove import sys from bottom call in test_idaklu_solver
- [x] Task 25: Skip the parameter_sets test
- [x] Task 26: Change the version requirements given the scikit-build-core docs
- [x] Task 28: Update docs for DAE solver (remove extra text from the tabs)


CHANGES IN PROGRESS:


COMPLETED, need reviews:
- [x] Task 4: Remove pins on other build-time dependencies like CasADi --- unpinned all instances of build-time dependencies, including `scikit-build-core` in the direction to remove editable-rebuilds, or at least give a warning to the users before using it.
- [x] Task 5: Remove `pip update` commands from `nox.`
- [x] Task 13: Remove `set CMAKE_GENERATOR="Visual Studio 17 2022"` and `set CMAKE_GENERATOR_PLATFORM=x64` and let CMake set it by itself.
- [x] Task 14: Set defaults for Windows for the remaining 4 environment variables in `CMakeLists.txt`. Use "if set, use that otherwise use default" logic for `VCPKG_ROOT` (rename `VCPKG_ROOT_DIR` to `VCPKG_ROOT`) --- Can not fix using CMake variables, have to set environment variables, hence wrote a batch script to set the env vars permanently for the Windows user. The CI still use pre-defined env vars because this avoids issues related to interactive scripts.
- [x] Task 19: Clean the `find BLAS` script
- [x] Task 3: Run tests, including IDAKLU ones, on all platforms (except doctests). --- random windows IDAKLU tests fail because of workers crashes. IREE tests had to be skipped because of attribute failure. Rest linux and macos tests are working.
- [x] Task 1: Check if the error in Doctests `(pybamm.parameter_sets)` can be resolved using an example instead of explicitly setting all the names. --- need to set all names
- [x] I unpinned `scikit-build-core` but will keep it as an option in the docs. I will write a warning saying that it is experimental.
    - [x] Task 2: Set the scikit-build-core pin to 0.10.5 from 0.10.3
    - [x] Task 6: Write a clearer message about editable-rebuild command, which is supposed to help developers using partial rebuilds (it was written to use that command to contribute to PyBaMM).
    - [x] Task 7: Remove the verbose flag from the editable-rebuilds command
- [x] Task 12: Set `python/python3` instead of `python3.X` in TOML kit installation in the guide.
- [x] Task 15: Remove ambiguity from `# in the PyBaMM/ directory`
- [x] Task 16: Migrate all Windows instructions to the existing `Install from source` doc.
- [x] Task 20: Add `pytest` documentation from Contributors guide in Develop branch.
- [x] Task 10: Include new "build" and "download_KLU_Sundials" directories in `.gitignore`
- [x] Task 18: Include `*.egg-info` pattern in `.gitignore`
- [x] Task 9: Update docs asking users to install VCPKG themselves.
    - [x] Task 8: Remove images from the installation guide.
    - [x] Task 11: Install VCPKG in a local directory
    - [x] Task 17: We do not need the MSVC image
- [x] Task 18: Unpin `graphviz` from docs and tests --- had to pin on tests because of `permission` errors, though it would work fine locally.
- [X] Task 21: Add IREE Support

cringeyburger commented:
With this commit, editable + partial rebuilds are available through `nox -s dev`. 
Note that we have to `export BUILD_IDAKLU=ON` to enable CMake and thus `IDAKLU` build

cringeyburger commented:
Updates till now:
1. Wheels are built properly, and the package can be installed simply with `pip install .[...]`.
2. CI Builds pass except for macOS, which can't find `Fortran` (which is actually installed before the build step) to find `SuiteSparse` libraries and `BLAS`. This shouldn't be a problem on local if `gfortran` is installed through `brew`.
3. `IDAKLU` is compiled properly and is installed during wheel builds and installation, given we have exported `BUILD_IDAKLU=ON` before running the build/install commands. Otherwise, a pure python package is installed.
4. Editable installs with rebuilds are working, and can be run using `pip install --no-build-isolation --config-settings=editable.rebuild=true -Cbuild-dir=build -ve.[...]`, provided we have the build-time dependencies, such as `cmake`, `pybind11` etc., are installed beforehand.
5. This is automated by running `nox -s dev` which installs the build-time dependencies. Note that this feature requires `tomlkit` to be installed before running any `nox` command. Hence it is better to install `nox` and `tomlkit` hand in hand.
6. It is to be noted that with editable installs and rebuilds, the `IDAKLU` solver would not work and show as `KLU not installed`. This is because `SuiteSparse` and `SUNDIALS` aren't interfaced with `pybind11` directly, and thus are not linked with the compiled artifact.
7. Docker builds are working, both locally and on CI. 
8. Doctests and other tests should work now since `nox` and `tomlkit` are installed together on CI.

cringeyburger commented:
Problem with Benchmarks:

1. The installation script installs the libs to `PyBaMM/sundials_KLU_libs`. 
2. CMake links them using `${CMAKE_CURRENT_SOURCE_DIR}/sundials_KLU_libs`.
3. The thing with ASV on CI is that the script responsible for installing the packages installs them in `/home/runner/work/PyBaMM/PyBaMM/sundials_KLU_libs/`.
4. But due to the build command: `python -m build --wheel -o {build_cache_dir} {build_dir}`, it makes a separate venv+pip environment, and the path to the libs is changed to `/home/runner/work/PyBaMM/PyBaMM/env/ec88222d46a18b5e9d12ba7194d09478/project/sundials_KLU_libs`

Because of this, CMake isn't finding the libs and is failing to build them. From what I know, we can try fixing this using `--no-build-isolation`. Let me know if this is problematic and if we don't want that.

cringeyburger commented:
Updates:
With this commit, the following things are covered:
1. Main migration to `scikit-build-core` with `IDAKLU` turned off as default. This is overridden on CI jobs and by exporting `BUILD_IDAKLU=ON`
2. Editable installs are now available with `nox -s dev`, which utilises `tomlkit` to install the build-time dependencies
3. CI builds for the builds are working for all three platforms
4. Docker builds are updated and working
5. Doctests, tests and benchmarks are working

cringeyburger commented:
Okay, now they work :)

kratman commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -125,6 +125,8 @@ jobs:
   run_example_tests:
     runs-on: ubuntu-latest
+    env:
+      BUILD_IDAKLU: ON
```
Can we control the idaklu build with something besides environment variables?

cringeyburger replied:
`scikit-build-core` provides support for [overrides](https://scikit-build-core.readthedocs.io/en/latest/overrides.html#any-matching-condition) based on environment variables. Until new features are added upstream, we have to deal with environment variables. Apart from this, we can have fallback options if the builds fail, but yes, they need to run first.

kratman replied:
Having to export does not seem like a great interface. I would prefer that it defaults to on and tries to recover if something is not found

kratman replied:
The reason I am saying this is that it would probably be good to move towards idaklu being the default solver

cringeyburger replied:
I see. I know that Windows tests and the Ubuntu doctests were failing because we haven't made compiling the IDAKLU solver optional and turned it off.
We can default it to on and track an environment variable to turn it off wherever it isn't needed. Would you like me to work on it?

cringeyburger replied:
One thing I noticed is that, during CI builds, even if `scikit-build-core` fallbacks to pure python build if the `IDAKLU` compilation fails, it still fails of `AssertionError: Metadata mismatch in WHEEL`. So, I don't think we can set a fall back, even for local development..? Please let me know if I am wrong

cringeyburger replied:
I tried running tests and building docs. They didn't fail and worked fine. It is because I have all the libs beforehand (also I use a `conda` shell, but the `nox` session builds worked fine). It would cause problems for people newly installing PyBaMM and others along the same line.

cringeyburger replied:
Just recreated a build failure locally, and yes, the fallback to pure python installation fails because of `AssertionError: Metadata mismatch in WHEEL`

cringeyburger commented:
After setting the compilation of the `IDAKLU` solver to default and providing a fallback, I ran tests, benchmarks, CI wheels' builds, and docker builds, and they all worked. 
Please note that I set `BUILD_IDAKLU=OFF` for tests where `IDAKLU` wasn't needed and fails the builds if set to ON.

cringeyburger commented:
After setting the compilation of the `IDAKLU` solver to default and providing a fallback, I ran workflows of tests, benchmarks, CI wheels' builds, and docker builds, and they all worked. 
Please note that I set `BUILD_IDAKLU=OFF` for tests where `IDAKLU` wasn't needed and fails the builds if set to ON.

cringeyburger commented:
I am not sure if the pre-commit's `blacken-docs` error relates to the changes I made.

cringeyburger commented:
The error (warning) in doctests is a quick fix: two links have the same name. I will change it as soon as all tests are okay.

cringeyburger commented:
Seems like there is a problem upstream for scipy==1.14.1, I am pinning it to 1.14.0

agriyakhetarpal commented:
It was the case of a missing M-series wheel for Python 3.12, which seems to have been fixed since the upload was processed manually. Could you unpin and check, @cringeyburger?

kratman commented on docs/source/user_guide/installation/install-source-windows.rst:
```diff
@@ -0,0 +1,382 @@
+
+.. code:: bash
+
+	choco install graphviz --version=8.0.5
```
Is this exact version required?

cringeyburger replied:
I am not sure myself. I copied this from the `run_periodic_tests.yml`. I thought it might have a linked issue.

kratman commented on docs/source/user_guide/installation/install-source-windows.rst:
```diff
@@ -0,0 +1,382 @@
+To install VCPKG, follow the steps below:
+
+1. Open a Command Shell, and navigate to ``C:\``.
+    * We are installing VCPKG in ``C:\vcpkg``
```
I would change the recommended directory to be something more local.

cringeyburger replied:
Usually VCPKG is installed in `C:\vcpkg`, `C:\Program Files\vcpkg` or `C:\Users\<username>\vcpkg`. All of these are good default system places.

kratman commented on docs/source/user_guide/installation/install-source-windows.rst:
```diff
@@ -0,0 +1,382 @@
+1. Open a Command Shell, and navigate to ``C:\``.
+    * We are installing VCPKG in ``C:\vcpkg``
+    * If you wish to install in a custom directory, change the environment variable ``VCPKG_ROOT_DIR`` before installation as mentioned herein :ref:`installing-pybamm`.
+2. Clone VCPKG's GitHub repository
+
+   .. code:: bash
+
+	  git clone https://github.com/microsoft/vcpkg.git
+
+
+   * If you do not have Git installed, download and install it from the `official Git download page <https://git-scm.com/download/win>`_.
+
+3. Navigate inside VCPKG directory and install it:
+
+   .. code:: bash
+
+	  cd vcpkg && bootstrap-vcpkg.bat
+
+4. Add VCPKG directory to ``Path`` environment variable:
+    * Copy the location ``C:\vcpkg`` or the location of your custom directory.
+    * In the start menu, search for ``Edit the system environment variables`` and open it.
+    * Under the ``Advanced`` tab, click on ``Environment Variables``
+        .. image:: images-windows-source/pic2-env-var-open.png
+			:scale: 70%
+
+
+    * Select ``Path`` from ``user variables`` if you want to set VCPKG just for you, or use ``Path`` from ``system variables`` to set VCPKG for all your system users.
+    	.. image:: images-windows-source/pic3-show-path.png
+		    :scale: 50%
+
+
+    * Select ``Path`` and select ``NEW``. Paste the location to VCPKG, then confirm your options by clicking ``OK``.
+    	.. image:: images-windows-source/pic4-save-env-var.png
+		  :height: 900px
+		  :width: 900px
+		  :scale: 70%
+
+
+5. Close and reopen Command Prompt and execute ``vcpkg`` to see an output of different options available for ``vcpkg``.
```
Can we instead tell them to follow the instructions on the vcpkg git repo? We don't want to have to change this if the recommendations from microsoft change.

cringeyburger replied:
Yes, we can do that. I will be removing this section. As you said, we can ask the users to install VCPKG and update their PATH permanently.

kratman commented on docs/source/user_guide/installation/install-source-windows.rst:
```diff
@@ -0,0 +1,382 @@
+
+.. code:: bash
+
+	  python3.X -m pip install --user tomlkit
```
It has been a while since I used windows, but I don't recall having to be so specific to the exact python version. I remember just `python` or `python3` working. Python 2 is not even supported anymore
```
	  python -m pip install --user tomlkit
```
or
```
	  python3 -m pip install --user tomlkit
```

cringeyburger replied:
You are right. I was under the presumption that the users had other lower Python versions. We can  edit this.

kratman commented on docs/source/user_guide/installation/install-source-windows.rst:
```diff
@@ -0,0 +1,382 @@
+		  set VCPKG_ROOT_DIR=C:\vcpkg
+		  set VCPKG_DEFAULT_TRIPLET=x64-windows-static-md
+		  set VCPKG_FEATURE_FLAGS=manifests,registries
+		  set CMAKE_GENERATOR="Visual Studio 17 2022"
+		  set CMAKE_GENERATOR_PLATFORM=x64
```
Do we need to set these? Usually CMake would try to figure out this sort of thing

cringeyburger replied:
Regarding the platform, I agree. We can remove that, and CMake can detect it. However, I was concerned about users with multiple Visual Studio versions.

agriyakhetarpal replied:
MSVC 2019 and 2022 are not ABI-compatible, but it's unlikely someone will have both, should be good to remove.

kratman commented on docs/source/user_guide/installation/install-source-windows.rst:
```diff
@@ -0,0 +1,382 @@
+
+   .. code-block:: bash
+
+		  set PYBAMM_USE_VCPKG=ON
+		  set VCPKG_ROOT_DIR=C:\vcpkg
+		  set VCPKG_DEFAULT_TRIPLET=x64-windows-static-md
+		  set VCPKG_FEATURE_FLAGS=manifests,registries
```
It would be nice if we could set these as part of the IDAKLU install script instead of making the user set them

cringeyburger replied:
We can do that by adding defaults to CMakeLists.txt. Apart from others, I will try to set `VCPKG_ROOT` (I will change VCPKG_ROOT_DIR to this) so we can use that if this is defined. Otherwise, we fall back to default.

cringeyburger replied:
I tried doing that but to no avail. The next best thing I could think of was writing a `.bat` script which can permanently set environment variables.

kratman commented on docs/source/user_guide/installation/install-source-windows.rst:
```diff
@@ -0,0 +1,382 @@
+
+.. code:: bash
+
+	  # in the PyBaMM/ directory
```
This could be ambiguous since there could be "PyBaMM" and "PyBaMM/src/pybamm"

kratman commented on docs/source/user_guide/installation/install-source-windows.rst:
```diff
@@ -0,0 +1,382 @@
+
+	pip install -ve .
+
+If you want to install to contribute to PyBaMM, use this command:
```
Same as above, this does not make sense.

kratman commented on docs/source/user_guide/installation/install-source-windows.rst:
Quite a lot of the windows instructions are the same as the other install from source instructions. Can we just add more OS dependent tabs to the main install from source page?

cringeyburger replied:
Sounds nice, I will give it a try.

kratman commented on docs/source/user_guide/installation/install-source-windows.rst:
```diff
@@ -0,0 +1,382 @@
+
+Note that you won't need Visual Studio 2022 entirely; just ``Desktop development with C++`` will suffice.
+
+.. image:: images-windows-source/pic1-msvc-select.png
+	:height: 456px
+	:width: 810px
+	:scale: 100%
+	:alt: Image with the packages to install
```
We probably don't need this image here

cringeyburger replied:
Okay, removing it.

agriyakhetarpal commented:
Thanks @kratman for reviewing the PR, especially for the documentation changes that have been added. Ankit is wrapping up his project and needs to write his final report for submission, so I'll try to find some time and prioritise reviewing this tomorrow and over the weekend.

On that note, @santacodes finished up his project today and is now looking to start with the IDAKLU solver separation milestone once he publishes his final report. @valentinsulzer and I were discussing with him some of the specifics around this in our meeting today, and whether it could be done in the course of a week or before we start with the v24.9 release cycle, which is not only subject to Santhosh's and Ankit's capabilities, but also to this PR getting merged, so the sooner it will be possible to, the better

kratman commented:
> On that note, @santacodes finished up his project today and is now looking to start with the IDAKLU solver separation milestone once he publishes his final report. @valentinsulzer and I were discussing with him some of the specifics around this in our meeting today, and whether it could be done in the course of a week or before we start with the v24.9 release cycle, which is not only subject to Santhosh's and Ankit's capabilities, but also to this PR getting merged, so the sooner it will be possible to, the better

Yeah that is why I was looking at this PR. To help with the separation task.

cringeyburger commented on src/pybamm/parameters/parameter_sets.py:
```diff
@@ -18,7 +18,7 @@ class ParameterSets(Mapping):
         >>> import pybamm
         >>> list(pybamm.parameter_sets)
-        ['Ai2020', 'Chen2020', ...]
+        ['Sulzer2019', 'Ai2020', 'Chen2020', 'Chen2020_composite', 'Ecker2015', 'Ecker2015_graphite_halfcell', 'Marquis2019', 'Mohtat2020', 'NCA_Kim2011', 'OKane2022', 'OKane2022_graphite_SiOx_halfcell', 'ORegan2022', 'Prada2013', 'Ramadass2004', 'Xu2019', 'ECM_Example', 'MSMR_Example']
```
The new test runner using `pytest` seems to need the complete output. Otherwise, it was failing the tests.

kratman replied:
I don't recall seeing failures in other PRs due to this. Is the failure something scikit-build-core specific?

cringeyburger replied:
I will try reverting the change. If it works with just an example, then that would be great. Otherwise, it might be related to scikit-build-core (I don't know how this is even related, to be honest).

agriyakhetarpal replied:
Yes, I had the same question, but I haven't figured out the problem. My hunch is that it changes the order of the files listed in the wheel metadata in comparison to the order set up in the `setuptools` build, causing `import pybamm` to change the order of what files are picked up first (and therefore this is seen in the doctests). Of course, this is only a hunch...

cringeyburger commented on scripts/Dockerfile:
```diff
@@ -28,11 +29,9 @@ RUN uv venv $VIRTUAL_ENV
 
-RUN uv pip install --upgrade setuptools wheel wget cmake
-
+RUN uv pip install --upgrade scikit-build-core==0.10.3 wget cmake casadi pybind11
```
Yes, scikit-build-core only provides experimental support for editable-rebuild installs.
1) To avoid unwanted breaking changes
2) Adding bounds helps with reproducibility in some aspects if someone were to build PyBaMM's older releases from source some time down the line.

Hence the necessity to pin scikit-build-core. For more info, have a look at the [documentation](https://scikit-build-core.readthedocs.io/en/stable/configuration.html#editable-installs).

cringeyburger replied:
Oh, I will update it. Thanks for pointing out!

cringeyburger commented on .github/workflows/test_on_push.yml:
```diff
@@ -34,6 +34,8 @@ jobs:
 
   run_unit_integration_and_coverage_tests:
     needs: style
+    env:
+      BUILD_IDAKLU: OFF
```
We don't install IDAKLU in tests because we don't install dependencies for Windows and Ubuntu doctests. To avoid this, we are skipping over IDAKLU tests in Windows in the develop branch. In this branch, we extended that list to Ubuntu and macOS.

Honestly, now that you say it, it was a dirty fix, and it's just better to skip the tests for Windows by setting `BUILD_IDAKLU=OFF` for it explicitly (otherwise, we need to install VCPKG packages and the tests would quickly go over 30+ minutes) and doctests (because we don't need IDAKLU over there anyways).

I will fix this part asap. Thanks for noticing!

kratman replied:
The problem with skipping the windows tests is that we won't know if there is a windows specific problem with IDAKLU. We want to make sure that the software works on all operating systems, we don't want to catch these problems only after we do a release....

Soon IDAKLU will be in its own repo and the process will get faster for PyBaMM, but until that time we are responsible for testing on windows as well.

cringeyburger replied:
Okay, I will make the tests run on all platforms

kratman replied:
Yeah it is fine to skip it something specific like a docs build test, but for IDAKLU and other core functionality it should run on all platforms

agriyakhetarpal replied:
Yes, I agree. I know we did discuss it at length about how it doesn't help with CI time, but now that we're going to unvendor the solver, the circumstances have changed. We can set `BUILD_IDAKLU` to `ON`, even if we need to take a hit with longer running CI jobs – since it's going to be temporary.

cringeyburger commented on pyproject.toml:
```diff
@@ -1,15 +1,15 @@
+    "scikit-build-core>=0.10,<=0.10.3",
     # On Windows, use the CasADi vcpkg registry and CMake bundled from MSVC
-    "casadi>=3.6.5; platform_system!='Windows'",
+    "casadi>=3.6.5,<=3.6.6; platform_system!='Windows'",
```
It was decided that it would be better to pin all build-time dependencies earlier. But we don't need to do that. Just pinning scikit-build-core is good. 

Fixing this!

cringeyburger commented on pyproject.toml:
```diff
@@ -1,15 +1,15 @@
 requires = [
-    "setuptools>=64",
-    "wheel",
+    "scikit-build-core>=0.10,<=0.10.3",
```
For the same reason, scikit-build-core only provides experimental support for editable-rebuild installs.
1. To avoid unwanted breaking changes
2. It can help reproducibility if someone builds PyBaMM's older releases from the source sometime soon.

kratman replied:
Does it work without editable rebuilds? I am bringing up the version pinning because it needs to be changed in multiple places whenever we update that pin. Ideally you should control this stuff in a single location or have a flexible range (e.g. `>=0.10.0,<0.11.0`).

At the moment the version needs to be manually updated in the pyproject file, the docker file, the documentation, and the CI.

cringeyburger replied:
Yes, you are right. It would be hard to keep track of the versions and update them. The normal editable installs (without rebuilds) work without any problem and don't depend on the `--no-build-isolation` flag too. 

We can remove the editable-rebuilds and keep only editable installation to remove the version pinning.

cringeyburger commented on noxfile.py:
```diff
@@ -218,10 +206,7 @@ def run_examples(session):
-    # https://bitbucket.org/pybtex-devs/pybtex/issues/169/replace-pkg_resources-with
-    # is fixed
-    session.install("setuptools", silent=False)
+    session.run("python", "-m", "pip", "install", "--upgrade", "pip")
```
I mean, why not? It is not that essential and can be removed if you like, but upgrading pip to be up-to-date with bug and security patches is good, and also ensures compatibility (some newer packages may need the latest version of pip).

kratman replied:
Good practice does not typically mean that we have to control it. The version of pip in the environment is the responsibility of the user.

In something like CI or a docker image, then it is our responsibility to make sure we are using a secure version, but updating the pip version each time tests are run is probably not the correct approach.

cringeyburger replied:
I see, I will the pip update commands

agriyakhetarpal replied:
> some newer packages may need the latest version of pip

This use case was valid, but it has largely diminished nowadays with the advent of the 2020 resolver and after pip 21.3+. Newer `pip` versions, even if they're a little outdated, can install almost all packages that play nicely with packaging tools and those that have precompiled wheels. It should be good to remove this command from all the sessions.

If you were wanting to do this to avoid the warning `pip` shows on the version being outdated, you can take a look at the `PIP_DISABLE_PIP_VERSION_CHECK` environment variable.

cringeyburger commented on docs/source/user_guide/installation/install-from-source.rst:
```diff
@@ -304,36 +359,53 @@ Here are some additional useful commands you can run with ``Nox``:
-Another common reason is that you forget to install a BLAS library such
-as OpenBLAS before installing SUNDIALS. Check the cmake output when you
-configured SUNDIALS, it might say:
+If you want to install to contribute to PyBaMM, use this command:
```
I meant that it would be nice for devs to opt for editable rebuilds over regular editable installations for "faster" development; at least, that was the idea. I should've been more clear here.

cringeyburger commented on docs/source/user_guide/installation/install-from-source.rst:
```diff
@@ -179,25 +197,62 @@ and run the tests to check your installation.
+
+.. code:: bash
+
+	pip install scikit-build-core==0.10.3 casadi cmake pybind11
```
Recappulating the above conversations:
1. editable rebuilds are experimental. Hence, version pinning is important to avoid breaking changes.
2. We can entirely remove editable-rebuilds and have normal editable installations. This would remove the need to pin the versions

cringeyburger commented on .gitignore:
```diff
@@ -85,6 +87,7 @@ sundials-*
 SuiteSparse-*
 build_sundials
 KLU_module_deps
+sundials_KLU_libs/include
```
The "build" dir is mentioned under "Sphinx build files" and all subdirs of "download_KLU_Sundials" are mentioned. Hence they both are ignored

cringeyburger commented on docs/source/user_guide/installation/install-from-source.rst:
```diff
@@ -179,25 +197,62 @@ and run the tests to check your installation.
 .. code:: bash
 
-	  pip install -e .[all,dev,docs]
+	  pip install --no-build-isolation --config-settings=editable.rebuild=true -ve .[all,dev,docs]
```
Okay, I will remove the flag if we decide to keep the editable-rebuilds. Otherwise, I will replace the editable-rebuilds with normal editable command.

cringeyburger commented on docs/source/user_guide/installation/images-windows-source/pic4-save-env-var.png:
If we ask the users to install VCPKG themselves, we don't have to put any images. I am in favour of this.

cringeyburger commented:
Due to the large number of conversations, it is becoming harder to track all the changes requested, changes in discussion and progress. 
To make it easier, I will be updating the first/top PR message after each update.

agriyakhetarpal commented on docs/source/user_guide/installation/install-source-windows.rst:
```diff
@@ -0,0 +1,382 @@
+.. code:: bash
+
+	  # in the PyBaMM/ directory
+	  python run-tests.py --unit
```
We removed the `run-tests.py` file in favour of `pytest` commands, could you change this? You may refer to them from the Contributing guide (from the `develop` branch).

cringeyburger commented:
@agriyakhetarpal Is [this](https://github.com/cringeyburger/PyBaMM/actions/runs/10536847176/job/29197492807#step:9:4315) supposed to happen? I tried running windows test with IDAKLU on.

agriyakhetarpal commented:
`test_symbol_visualise` can be skipped, could you add a pytest condition there to skip it if the `latex` program is not found? You can use `shutil.whicj()` to find it. As for the IDAKLU-JAX test, I'm not too surprised, because we haven't run the tests against the solver on PRs before – I knew there would be some failures coming up. I would suggest running them again, and if they keep failing, we can mark them with `pytest.mark.xfail` and fix them later, since they aren't related to the build system migration.

cringeyburger commented:
@agriyakhetarpal Check [this](https://github.com/cringeyburger/PyBaMM/actions/runs/10538824663/job/29201752124#step:8:6675) too. IREE is only for MacOS 13 and higher + python 3.11, and it is the only failing test.

cringeyburger commented:
From this job, it is known that the IDAKLU tests failing on windows are highly inconsistent. Different tests fail at different runs. It would be better to skip the IDAKLU build on Windows for now and tackle it later. 

As for IREE, they were consistent so I wrote conditionals with xfail

cringeyburger commented:
@agriyakhetarpal @kratman, with this commit, all the requested changes have been made. It would be great if you could review this PR again.

Please note that before merging, I need to change the `git clone` to upstream in Dockerfile and benchmarks (and maybe some other files).

cringeyburger commented:
Hmm... IREE tests are passing in CI builds but failing in unit tests. I have to take a look at this.

cringeyburger commented:
The build failure is due to some network problem (don't know how). It will work if run again.

kratman commented on docs/source/user_guide/installation/install-from-source.rst:
```diff
@@ -237,31 +486,32 @@ Finally, to run the unit and the integration suites sequentially, use
 
-Using the test runner
-~~~~~~~~~~~~~~~~~~~~~~
+Using ``pytest``
```
Just a note: this is fine the way it was, pytest is still a test runner

kratman reviewed (changes requested):
Partial review. I still need to go over the docs in more detail

cringeyburger commented on .github/workflows/publish_pypi.yml:
```diff
@@ -288,7 +288,10 @@ jobs:
 
   publish_pypi:
     # This job is only of value to PyBaMM and would always be skipped in forks
-    if: github.event_name != 'schedule' && github.repository == 'pybamm-team/PyBaMM'
+    if: >-
+      (github.event_name == 'workflow_dispatch' ||
+      github.event_name == 'release') &&
+      github.repository == 'pybamm-team/PyBaMM'
```
If I am not wrong, it was made by @agriyakhetarpal to make pushing to PYPI harder

cringeyburger commented on .github/workflows/test_on_push.yml:
```diff
@@ -32,37 +33,58 @@ jobs:
     needs: style
-    runs-on: ${{ matrix.os }}
+    env:
+      BUILD_IDAKLU: OFF
```
Have a look this [job](https://github.com/cringeyburger/PyBaMM/actions/runs/10536847176/job/29197492807#step:9:3760). I turned on IDAKLU build for Windows. There was no problem with building it, but the tests failed because of the workers crashing. I tried the same on my Windows laptop, and it's even worse.

Even the current development branch doesn't test for IDAKLU on Windows, so I think we might have to spend a bit more time, and guessing by the amount of work needed, it's better tackled on another fresh PR.

kratman replied:
Idaklu-jax had an issue that was fixed here #4379 Not sure if that was related.

cringeyburger replied:
Just enabled IDAKLU in Windows tests

cringeyburger replied:
Hmm.. There is a problem with the repair step of the Windows CI builds now, caused by the recent updates in the development branch. Here is the [job](https://github.com/cringeyburger/PyBaMM/actions/runs/10567902406/job/29277837156#step:8:748)

cringeyburger replied:
Am I expected to fix this? Looks hard 😅

cringeyburger replied:
I was trying to build pybamm locally on windows, and to do that I had to update MSVC 2022. So is it related to that?

cringeyburger replied:
@agriyakhetarpal Also for the single thread tests, yes it still failed

cringeyburger commented on .gitignore:
```diff
@@ -44,6 +44,9 @@ input/*
 # keep images required by notebooks
 !docs/examples/notebooks/creating_models/SEI.png
 
+# keep images required for installation guide
+!docs/source/user_guide/installation/images-windows-source/*.png
+
```
Yes! I forgot to remove this. Thank you

cringeyburger commented on tests/unit/test_solvers/test_idaklu_solver.py:
```diff
@@ -7,6 +7,7 @@
 import unittest
 import pytest
 import numpy as np
+import sys
```
Okay noted.

cringeyburger commented on CMakeLists.txt:
```diff
@@ -159,6 +173,21 @@ endif()
 include_directories(${SuiteSparse_INCLUDE_DIRS})
 target_link_libraries(idaklu PRIVATE ${SuiteSparse_LIBRARIES})
 
+if(DEFINED ENV{CIBUILDWHEEL} AND CMAKE_SYSTEM_NAME STREQUAL "Linux")
+  set(SUNDIALS_LIBRARY_DIR "${INSTALL_DIR}/lib64")
+  set(SUITESPARSE_LIBRARY_DIR "${INSTALL_DIR}/lib64")
+else()
+  set(SUNDIALS_LIBRARY_DIR "${INSTALL_DIR}/lib")
+  set(SUITESPARSE_LIBRARY_DIR "${INSTALL_DIR}/lib")
+endif()
```
The SUNDIALS libs for Linux CI are unpacked in lib64 format, so it was necessary for this.

cringeyburger commented on docs/source/user_guide/installation/install-from-source.rst:
```diff
@@ -12,146 +12,288 @@ Here is a gentle introduction/refresher: `Python Virtual Environments: A Primer
+	.. code::
 
-Note that running ``nox`` will create new virtual environments for you to use, so you do not need to create one yourself.
+		C:\Users\<USERNAME>\AppData\Roaming\Python\Python3<X>\Scripts
```
Yes, this is needed. I tested it locally and without this nox wouldn't work

agriyakhetarpal replied:
Could we instruct users to use `python -m nox` for this? That should work better instead of asking themselves to update `PATH`.

cringeyburger commented on scripts/windows_setup.bat:
```diff
@@ -0,0 +1,10 @@
+@echo off
+
+echo Setting environment variables...
+
+setx PYBAMM_USE_VCPKG ON
+setx VCPKG_DEFAULT_TRIPLET x64-windows-static-md
+setx VCPKG_FEATURE_FLAGS manifests,registries
+
+echo Environment variables have been set. Please open a new Command Prompt to see the changes.
+pause
```
Okay, I will try this and update you

agriyakhetarpal replied:
`install_KLU_Sundials.py` isn't supported and does not work for Windows, it's a script made just for Unix-based platforms

cringeyburger replied:
I wanted to clarify: currently we don't support windows users running the `install_KLU_Sundials.py` script. Adding this feature in this script would change that, and prompting more changes and maybe errors. 

I think the way it is set currently is better as it is: 
1. `install_KLU_Sundials` for linux and macos to install the libs
2. `setup_windows.bat` to setup the env vars and the installation is carried out automatically by vcpkg.

I feel this is better. What do you say?

agriyakhetarpal replied:
Yes, I feel this is better for now. We should start testing `setup_windows.bat` in CI sometime soon, though. 

A cross-platform script that handles everything from Windows to Linux would be great, but maybe we would need to switch to `vcpkg` entirely for that.

cringeyburger replied:
I did try the script locally and it works. For CI though, because of the `pause` in the last line of the script, the CI fails. From what I understand, it fails to work with interactive scripts (pause is needed to get that "press any key to continue" screen for the users, otherwise it would just flash and close). So I think it would be nice that we work with env vars like we have been, or find a way to make the CI interactive (??).

kratman replied:
Interactive is terrible for CI. What is the pause needed for?

cringeyburger replied:
pause literally pauses the script till any key is pressed to close it. It's more of a UI/UX thing for users. Otherwise when you run it, the environment vars would be set and the terminal closed automatically

agriyakhetarpal replied:
CI can't be made interactive, it's non-interactive by design :) Unless you connect to a runner via SSH. Not needed for now, but once we add it to CI somewhere, you can check for `CIBUILDWHEEL` as an environment variable and only add the `pause` there – we do something similar for the `docs` session in `noxfile.py`.

cringeyburger commented on FindSuiteSparse.cmake:
```diff
@@ -37,6 +37,20 @@
 #   system paths.
 #
 
+if(NOT CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+  enable_language(Fortran)
+  set(IDAKLU_BLAS_VENDOR
+    "OpenBLAS"
+    CACHE
+      STRING
+      "Sets the BLAS/LAPACK vendor. See https://cmake.org/cmake/help/latest/module/FindBLAS.html#blas-lapack-vendors."
+  )
+  set(BLA_VENDOR ${IDAKLU_BLAS_VENDOR})
+
+  # openBLAS: include a "lib" prefix in its names
+  set(CMAKE_FIND_LIBRARY_PREFIXES "" lib)
+endif()
```
As explained by @agriyakhetarpal,
>We currently distribute the IDAKLU solver in the conda-forge distribution for PyBaMM, and if we would later in the future, it's going to be through another feedstock, which probably requires different configuration.

>We don't currently support linking against MKL, and I'm not sure if we will – that would just invite licensing troubles

cringeyburger replied:
Also, we are suggesting users install Openblas (and give commands to install it), so I think the probability of someone with other BLAS is low...

cringeyburger commented on docs/source/user_guide/installation/install-from-source.rst:
```diff
@@ -12,146 +12,288 @@ Here is a gentle introduction/refresher: `Python Virtual Environments: A Primer
+	To install PyBaMM, you will need:
 
-.. code:: bash
+	- Python 3 (PyBaMM supports versions 3.9, 3.10, 3.11, and 3.12)
+	- The Python headers file for your current Python version.
+	- A BLAS library (for instance `openblas <https://www.openblas.net/>`_).
+	- A C compiler (ex: ``gcc``).
+	- A Fortran compiler (ex: ``gfortran``).
+	- ``graphviz`` (optional), if you wish to build the documentation locally.
+	- ``pandoc`` (optional) to convert the example Jupyter notebooks when building the documentation.
+	- ``texlive-latex-extra`` (optional) to convert model equations in latex.
+	- ``dvipng`` (optional) to convert a DVI file to a PNG image.
```
For this section, the steps differ a bit too much for Windows, so I had to set all three together on top

cringeyburger commented on docs/source/user_guide/installation/install-from-source.rst:
```diff
@@ -12,146 +12,288 @@ Here is a gentle introduction/refresher: `Python Virtual Environments: A Primer
+
+.. tab:: Windows
+
+	PyBaMM comes with a DAE solver based on the IDA solver provided by the SUNDIALS library. To use this solver, you must ensure you have the necessary SUNDIALS components installed on your system.
```
You are right in this case, the text is common for both. I will fix this

cringeyburger commented on pyproject.toml:
```diff
@@ -156,15 +156,23 @@ Xu2019 = "pybamm.input.parameters.lithium_ion.Xu2019:get_parameter_values"
-[tool.setuptools]
-include-package-data = true
+[tool.scikit-build]
+minimum-version = "0.10"
```
Okay, the documentation for scikit-build-core changed. It seems that we don't need to set it like this anymore (before we had to). I will fix this

agriyakhetarpal commented on CMakeLists.txt:
```diff
@@ -108,7 +117,9 @@ pybind11_add_module(idaklu
 )
 
-if (NOT DEFINED USE_PYTHON_CASADI)
+if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
+  set(USE_PYTHON_CASADI FALSE)
+else()
```
Yes. This behaviour was previously handled using an environment variable `USE_PYTHON_CASADI`, which converted it to a CMake option `-DUSE_PYTHON_CASADI` that was passed to the CMake command in a subprocess inside `setup.py`. We now set it directly here because we no longer have a `setup.py` file.

agriyakhetarpal commented on scripts/install_KLU_Sundials.py:
```diff
@@ -285,7 +285,7 @@ def parallel_download(urls, download_dir):
 
 # Create download directory in PyBaMM dir
 pybamm_dir = os.path.split(os.path.abspath(os.path.dirname(__file__)))[0]
-download_dir = os.path.join(pybamm_dir, "install_KLU_Sundials")
+download_dir = os.path.join(pybamm_dir, "download_KLU_Sundials")
```
The Makefiles for the libraries are still placed in subfolders inside `download_KLU_Sundials`, so this looks like it was just a rename of the folder. The object files are installed in `sundials_KLU_libs/`.

cringeyburger replied:
The way it's set is that downloaded parts will be in the project dir, but the installation of the libs can be set to any other custom directory (also that `install_KLU_Sundials` is not a good name for a dir containing only downloaded parts)

agriyakhetarpal commented:
I triggered https://github.com/pybamm-team/PyBaMM/actions/runs/10577390897/job/29305220084 from the `develop` branch, let's see if the Windows wheel repair step fails there, too.

Also, @kratman reported some troubles when installing PyBaMM from this branch on macOS. In his words, "It looks like it cannot find whatever library defines 'realtype' and 'booleantype'"

I'm not sure about that because I don't have the same issue on my macOS machine because everything compiles successfully after `nox -s pybamm-requires` and `nox -s dev`, but I do have a different issue though – the `idaklu` shared object isn't placed in the `src/pybamm/solvers/` directory after installation for some reason (I did verify by changing the current `pybamm/solvers/` to `src/pybamm/solvers/` in `CMakeLists.txt` but to no avail). The shared object is placed in `build/{wheel_tag}`, but not copied into the PyBaMM directory (therefore the IDAKLU solver tests are skipped). I am on an M-series (ARM) device if that helps.

cringeyburger commented:
Okay, I found out that either doing `pip install ...` or `nox -s dev` installed the solver in site packages, somewhere here `/home/yukinatsu/PyBaMM/venv/lib/python3.11/site-packages/pybamm/solvers/idaklu.cpython-311-x86_64-linux-gnu.so`, which is different for different virtual envs. 
I also remember the exact moment I changed this (I was trying to fix the BLAS not found problem, so I was trying different installation sites, then I wrote the BLAS script in `FindSuiteSparse.cmake` and forgot about it).

I fixed it with the following commit, now it installs within the src/pybamm/solvers. I also tested using `pytest -m unit` and `nox -s unit` and they worked like they before.

------

I don't know how IDAKLU tests were skipped for you.

Even without the solver being installed in `src/pybamm/solvers`, running `pip install ...` with `pytest -m unit` will work perfectly. And `nox -s dev` followed by, or just `nox -s unit` will work perfectly.

It will skip tests if you did `nox -s dev` and tried to do `pytest -m unit` without being in the virtual environment created by `nox`, since pybamm isn't installed locally (also note that pybamm might be installed in your local/conda env from previous attempts, though not perfectly).

It will work if you are in venv virtual environment (created during `nox -s dev`) and did `pytest -m unit` (tested it). 

Under a self-made virtual environment (name it as venv inside the project root because if `nox` commands are later run, it will use the same env), using manual commands like `pip install .[all,dev,jax]` or `pip install -e .[all,dev,jax]` followed by `pytest -m unit` works. 

`nox -s pybamm-requires` is common for both venv and self-made env, as it just downloads and installs in the project root dir.

NOTE: Remove `build` before running commands to get full on fresh compilation

agriyakhetarpal commented:
I just confirmed that the wheel builds and repairs on macOS locally with `BUILD_IDAKLU="ON" python -m build --wheel --no-isolation` and then `delocate-wheel -v dist/*whl` are fine when all of the libs are stored in `sundials_KLU_libs/`, which is great.

The problem is somewhere with the editable installation, and the command

```
BUILD_IDAKLU="ON" pip install --no-build-isolation -Cbuild-dir=build -ve.
```

copies the generated shared object (`idaklu.cpython-312-darwin.so`) for me in `PyBaMM/venv/lib/python3.12/site-packages/pybamm/solvers/`, and not in `PyBaMM/src/pybamm/solvers` – which is why it isn't found. This is the case for both scikit-build-core 0.9.10 and 0.10.5, so it either isn't an upstream issue, or it is but hasn't been discovered, or we are using an incorrect mechanism to load the shared object in `pybamm.have_idaklu()`.

Non-editable installations (without `-e` flag) work fine.

cringeyburger commented:
Also, I noticed that the repair step in the develop branch with Windows CI failed like it did with mine.

cringeyburger commented:
Note that, with this commit, the wheel built locally won't have the solver in it, since no env vars are set during local wheel builds, so delocate will fail

agriyakhetarpal commented:
That doesn't sound like the best way to fix it. We would want to handle both local wheel builds and repairs, and also for `cibuildwheel` at the same time, so this is more of a workaround that we can spend some time fixing in a more robust manner.

agriyakhetarpal commented:
The Windows wheel repair failure is indeed not related, I posted about it in https://github.com/pybamm-team/PyBaMM/issues/4385#issuecomment-2312564419

agriyakhetarpal commented:
Could you try out `build` instead of `pip` for building the wheels without your last commit? i.e., by changing to `CIBW_BUILD_FRONTEND: "build; args: --no-isolation"`?

agriyakhetarpal commented:
I guess we should go ahead with your solution, then – and revisit once we have found where the problem is?

cringeyburger commented:
Yes, that would be great. I will try to figure something out by the end

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -26,27 +26,72 @@ concurrency:
+          python -m pip install pre-commit
+          pre-commit run -a
+
+  run_windows_unit_integration_tests:
+    needs: style
+    env:
+      BUILD_IDAKLU: OFF
```
This should be `ON`?

```suggestion
  run_windows_unit_integration_tests:
    needs: style
    env:
      BUILD_IDAKLU: ON
```

cringeyburger replied:
Yes, I haven't touched periodic tests like we did tests on push. I will fix it

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -56,46 +101,110 @@ jobs:
+      - name: Check out PyBaMM repository
+        uses: actions/checkout@v4
+
+      - name: Install Linux system dependencies
+        uses: awalsh128/cache-apt-pkgs-action@v1.4.2
+        with:
+          packages: gfortran gcc graphviz pandoc
+          execute_install_scripts: true
```
You can replace these with a simple `apt-get install` in the step below, because we don't need to cache anything for scheduled tests.

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -56,46 +101,110 @@ jobs:
+      - name: Install nox and tomlkit
+        run: python -m pip install nox tomlkit
+
+      - name: Cache pybamm-requires nox environment for GNU/Linux
+        uses: actions/cache@v4
+        with:
+          path: |
+            # Headers and dynamic library files for SuiteSparse and SUNDIALS
+            ${{ github.workspace }}/install_KLU_Sundials/
+          key: nox-ubuntu-latest-pybamm-requires-${{ steps.setup-python.outputs.python-version }}-${{ hashFiles('**/install_KLU_Sundials.py', '**/noxfile.py', '**/test_on_push.yml') }}
```
```suggestion
```

This is not needed for the scheduled tests

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -105,17 +214,27 @@ jobs:
 
       - name: Install Linux system dependencies
+        uses: awalsh128/cache-apt-pkgs-action@v1.4.2
+        with:
+          packages: graphviz pandoc
+          execute_install_scripts: true
+
+      # dot -c is for registering graphviz fonts and plugins
+      - name: Install TeXLive for Linux
         run: |
           sudo apt-get update
-          sudo apt-get install graphviz pandoc texlive-latex-extra dvipng
+          sudo dot -c
+          sudo apt-get install texlive-latex-extra dvipng
```
Same suggestion as above

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -105,17 +214,27 @@ jobs:
+          sudo apt-get install texlive-latex-extra dvipng
 
       - name: Set up Python
+        id: setup-python
```
```suggestion
```

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -105,17 +214,27 @@ jobs:
         uses: actions/setup-python@v5
         with:
           python-version: 3.11
+          cache: 'pip'
```
```suggestion
```

We want the scheduled tests to run without caching

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -124,25 +243,46 @@ jobs:
 
       - name: Install Linux system dependencies
+        uses: awalsh128/cache-apt-pkgs-action@v1.4.2
+        with:
+          packages: gfortran gcc graphviz pandoc
+          execute_install_scripts: true
+
+      # dot -c is for registering graphviz fonts and plugins
+      - name: Install OpenBLAS and TeXLive for Linux
         run: |
           sudo apt-get update
-          sudo apt-get install gfortran gcc graphviz pandoc libopenblas-dev texlive-latex-extra dvipng
+          sudo dot -c
+          sudo apt-get install libopenblas-dev texlive-latex-extra dvipng
```
Same suggestion as above

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -124,25 +243,46 @@ jobs:
+          sudo apt-get install libopenblas-dev texlive-latex-extra dvipng
 
       - name: Set up Python 3.12
+        id: setup-python
```
```suggestion
```

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -124,25 +243,46 @@ jobs:
         uses: actions/setup-python@v5
         with:
           python-version: 3.12
+          cache: 'pip'
```
```suggestion
```

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -124,25 +243,46 @@ jobs:
+      - name: Install nox and tomlkit
+        run: python -m pip install nox tomlkit
+
+      - name: Cache pybamm-requires nox environment for GNU/Linux
+        uses: actions/cache@v4
+        with:
+          path: |
+            # Headers and dynamic library files for SuiteSparse and SUNDIALS
+            ${{ github.workspace }}/sundials_KLU_libs/
+          key: nox-ubuntu-latest-pybamm-requires-${{ steps.setup-python.outputs.python-version }}-${{ hashFiles('**/install_KLU_Sundials.py', '**/noxfile.py', '**/test_on_push.yml') }}
```
```suggestion
```

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -152,25 +292,46 @@ jobs:
         uses: actions/checkout@v4
 
       - name: Install Linux system dependencies
+        uses: awalsh128/cache-apt-pkgs-action@v1.4.2
+        with:
+          packages: gfortran gcc graphviz
+          execute_install_scripts: true
+
+      # dot -c is for registering graphviz fonts and plugins
+      - name: Install OpenBLAS and TeXLive for Linux
```
Same suggestion as above

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -152,25 +292,46 @@ jobs:
+          sudo apt-get install libopenblas-dev texlive-latex-extra dvipng
 
       - name: Set up Python 3.12
+        id: setup-python
```
```suggestion
```

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -152,25 +292,46 @@ jobs:
         uses: actions/setup-python@v5
         with:
           python-version: 3.12
+          cache: 'pip'
```
```suggestion
```

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -152,25 +292,46 @@ jobs:
+      - name: Install nox and tomlkit
+        run: python -m pip install nox tomlkit
+
+      - name: Cache pybamm-requires nox environment for GNU/Linux
+        uses: actions/cache@v4
+        with:
+          path: |
+            # Headers and dynamic library files for SuiteSparse and SUNDIALS
+            ${{ github.workspace }}/sundials_KLU_libs/
+          key: nox-${{ matrix.os }}-pybamm-requires-${{ steps.setup-python.outputs.python-version }}-${{ hashFiles('**/install_KLU_Sundials.py', '**/noxfile.py', '**/test_on_push.yml') }}
```
```suggestion
```

agriyakhetarpal commented on .github/workflows/test_on_push.yml:
```diff
@@ -1,3 +1,4 @@
+
 name: PyBaMM
```
```suggestion
name: PyBaMM
```

agriyakhetarpal commented on .github/workflows/test_on_push.yml:
```diff
@@ -32,37 +33,78 @@ jobs:
-        os: [ubuntu-latest, macos-12, macos-14, windows-latest]
         python-version: ["3.9", "3.10", "3.11", "3.12"]
-    name: Tests (${{ matrix.os }} / Python ${{ matrix.python-version }})
+    name: Tests (windows_latest / Python ${{ matrix.python-version }})
```
```suggestion
    name: Tests (windows-latest / Python ${{ matrix.python-version }})
```

agriyakhetarpal commented on .github/workflows/test_on_push.yml:
```diff
@@ -32,37 +33,78 @@ jobs:
+  run_macos_unit_integration_tests:
+    needs: style
+    runs-on: ${{ matrix.os }}
+    env:
+        PYBAMM_IDAKLU_EXPR_CASADI: ON
+        PYBAMM_IDAKLU_EXPR_IREE: ON
```
```suggestion
    env:
      PYBAMM_IDAKLU_EXPR_CASADI: ON
      PYBAMM_IDAKLU_EXPR_IREE: ON
```

agriyakhetarpal commented on .github/workflows/test_on_push.yml:
```diff
@@ -72,61 +114,107 @@ jobs:
         run: |
           brew analytics off
-          brew install graphviz libomp
+          brew install graphviz libomp openblas
```
```suggestion
          brew install graphviz libomp
```

Same question as above

cringeyburger replied:
We probably don't need it. I will test it and let you know

cringeyburger replied:
Yes we don't need it. I already removed it from the test workflows and they pass

agriyakhetarpal commented on .github/workflows/test_on_push.yml:
```diff
@@ -72,61 +114,107 @@ jobs:
         # sometimes gfortran cannot be found, so reinstall gcc just to be sure
         run: |
           brew analytics off
-          brew install graphviz libomp
+          brew install graphviz libomp openblas
           brew reinstall gcc
 
-      - name: Install Windows system dependencies
-        if: matrix.os == 'windows-latest'
-        run: choco install graphviz --version=8.0.5
-
       - name: Set up Python ${{ matrix.python-version }}
         id: setup-python
         uses: actions/setup-python@v5
         with:
           python-version: ${{ matrix.python-version }}
           cache: 'pip'
 
-      - name: Install nox
-        run: python -m pip install nox
+      - name: Install nox and tomlkit
+        run: python -m pip install nox tomlkit
 
-      - name: Cache pybamm-requires nox environment for GNU/Linux and macOS
+      - name: Cache pybamm-requires nox environment for macOS
         uses: actions/cache@v4
-        if: matrix.os != 'windows-latest'
         with:
           path: |
-            # Repository files
-            ${{ github.workspace }}/pybind11/
-            ${{ github.workspace }}/install_KLU_Sundials/
             # Headers and dynamic library files for SuiteSparse and SUNDIALS
-            ${{ env.HOME }}/.local/lib/
-            ${{ env.HOME }}/.local/include/
+            ${{ github.workspace }}/install_KLU_Sundials/
```
You can also cache `sundials_KLU_libs/` here and possibly escape the installation step.

agriyakhetarpal commented on .github/workflows/test_on_push.yml:
```diff
@@ -198,20 +286,16 @@ jobs:
-            ${{ env.HOME }}/.local/lib/
-            ${{ env.HOME }}/.local/include/
-          key: nox-${{ matrix.os }}-pybamm-requires-${{ steps.setup-python.outputs.python-version }}-${{ hashFiles('**/install_KLU_Sundials.py', '**/noxfile.py', '**/test_on_push.yml') }}
+            ${{ github.workspace }}/sundials_KLU_libs/
```
Here, you cached `sundials_KLU_libs/` – I would suggest caching both folders and maintaining consistency for all Intel and macOS jobs.

agriyakhetarpal commented on .github/workflows/update_version.yml:
```diff
@@ -48,7 +48,7 @@ jobs:
       - name: Install dependencies
         run: |
-          pip install wheel
+          pip install scikit-build-core
```
```suggestion
```

agriyakhetarpal replied:
`scikit-build-core` and other build-time dependencies will be pulled in with the installation step below this line.

agriyakhetarpal commented on CMakeLists.txt:
```diff
@@ -35,11 +35,17 @@ add_compile_definitions(_GLIBCXX_USE_CXX11_ABI=0)
 if(NOT PYBIND11_DIR)
   set(PYBIND11_DIR pybind11)
 endif()
```
```suggestion
```

We can remove these lines because we don't need to git-clone `pybind11` anymore.

agriyakhetarpal commented on noxfile.py:
```diff
@@ -232,8 +210,20 @@ def set_dev(session):
+    else:
+        python = os.fsdecode(VENV_DIR.joinpath("bin/python"))
     components = ["all", "dev", "jax"]
+    with open("pyproject.toml", encoding="utf-8") as file:
+        toml_data = tomlkit.parse(file.read())
+    build_dependencies = toml_data.get("build-system").get("requires")
+    if sys.platform == "win32":
+        build_dependencies = [
+            dep
+            for dep in build_dependencies
+            if not (dep.startswith("casadi") or dep.startswith("cmake"))
+        ]
```
Could you convert this to a reusable function at the top of the file and use it for all the sessions? All of the sessions should install the project without build isolation, not just `nox -s dev`. Also, we should drop `tomlkit` entirely from all workflows, see the `nox` documentation here on how to replace it –

https://github.com/wntrblm/nox/blob/e0868b6c092c80837f86463a54d34379851d43d7/docs/tutorial.rst#loading-dependencies-from-pyprojecttoml-or-scripts

cringeyburger replied:
Okay, working on it!

agriyakhetarpal commented:
Also, it would be helpful if you could [xfail](https://docs.pytest.org/en/stable/how-to/skipping.html#xfail-mark-test-functions-as-expected-to-fail) the IDAKLU-Jax tests if on Windows to get CI green. I think this is almost ready.

cringeyburger commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -56,46 +101,110 @@ jobs:
         run: |
           brew analytics off
-          brew install graphviz libomp
+          brew install graphviz libomp openblas
```
Hmm, it should work since we are not using the BLAS script for macOS.

cringeyburger commented:
The Windows CI tests fail at random IDAKLU-Jax tests. They are inconsistent with each Python version and not even with each job. I think skipping IDAKLU-Jax tests for Windows is better than xfail-ing them.

agriyakhetarpal commented on docs/source/user_guide/installation/install-from-source.rst:
```diff
@@ -12,146 +12,276 @@ Here is a gentle introduction/refresher: `Python Virtual Environments: A Primer
+
+	   .. code::
+
+			.\Scripts\windows_setup.bat
```
```suggestion
			.\scripts\windows_setup.bat
```

agriyakhetarpal commented on docs/source/user_guide/installation/install-from-source.rst:
```diff
@@ -179,32 +309,139 @@ and run the tests to check your installation.
 Running the tests
 -----------------
 
+.. tab:: GNU/Linux
+
+	Make sure to install ``texlive-latex-extra`` to pass all tests. Otherwise, you can safely ignore the failed tests needing it.
+
```
It would be better if we skip the test conditionally based on whether `latex` is installed (run the test only if `latex --version` exits with no errors and returns an exit code of 0). We can remove this line after that.

agriyakhetarpal commented on docs/source/user_guide/installation/install-from-source.rst:
```diff
@@ -304,36 +542,86 @@ Here are some additional useful commands you can run with ``Nox``:
+
+	.. code:: bash
+
+		pip install -e .
```
```suggestion
		pip install -e . --no-build-isolation
```

i.e., we should always recommend installation without isolation

agriyakhetarpal commented on noxfile.py:
```diff
@@ -52,9 +52,27 @@ def set_iree_state():
+    build_dependencies = nox.project.load_toml("pyproject.toml")["build-system"][
+        "requires"
+    ]
+    if sys.platform == "win32":
+        build_dependencies = [
+            dep
+            for dep in build_dependencies
+            if not (dep.startswith("casadi") or dep.startswith("cmake"))
+        ]
```
Question: Do we need to process the list of dependencies beforehand? The installation using `session.install("<...>")` should work equally well with [environment markers](https://packaging.python.org/en/latest/specifications/dependency-specifiers/#environment-markers).

cringeyburger replied:
Yes, I tested it out, and the env markers work well without the pre-processing. I am removing it.

cringeyburger commented:
How do I deal with the codacy analysis? I don't understand what it means..

kratman commented:
> How do I deal with the codacy analysis? I don't understand what it means..

It is telling you that it is best not to check for latex in that manner. Basically when you do a system call, it will run whatever you put in the command. Theoretically, you could make any application be named "latex" and that line would run it. Sometimes things like subprocess are needed, but codacy is telling you to check if it is actually needed in the tests.

On a similar note, why is this extra checking for latex needed? This PR is supposed to be about the build system, but that seems a bit out of scope.

>  I don't understand what it means..

Not sure if you know this, but if you click on the codacy message, it opens a window that gives a detailed explanation of what it means and how you can fix it.

kratman commented on tests/unit/test_solvers/test_idaklu_jax.py:
```diff
@@ -100,6 +102,10 @@ def test_instantiate_fails(self):
     not pybamm.have_idaklu() or not pybamm.have_jax(),
     "IDAKLU Solver and/or JAX are not available",
 )
+@pytest.mark.skipif(
+    sys.platform.startswith("win"),
+    reason="Problem with tests crashing workers on Windows",
+)
 class TestIDAKLUJax(unittest.TestCase):
```
Why wasn't this needed before?

cringeyburger replied:
It was, but we never added it. 
>Have a look this [job](https://github.com/cringeyburger/PyBaMM/actions/runs/10536847176/job/29197492807#step:9:3760). I turned on IDAKLU build for Windows. There was no problem with building it, but the tests failed because of the workers crashing. I tried the same on my Windows laptop, and it's even worse.
Even the current development branch doesn't test for IDAKLU on Windows, so I think we might have to spend a bit more time, and guessing by the amount of work needed, it's better tackled on another fresh PR.

Also
>Not sure if you know this, but if you click on the codacy message, it opens a window that gives a detailed explanation of what it means and how you can fix it.

I know about this and checked it too, but it was kinda vague 😅.

kratman replied:
If the skip test was needed before, then why weren't the failing tests showing up in CI before we used sbc? That is what I am trying to understand. What changed

cringeyburger commented:
It seems that installing openBLAS through `cache-apt-pkgs-action` isn't working. We need to install it separately.

agriyakhetarpal commented:
Yes, we used to install TeXLive and OpenBLAS separately earlier

agriyakhetarpal commented:
Now that we are done with the v24.9 release, let's plan to merge this in the coming week.

agriyakhetarpal commented:
I've been trying to get to this PR but it has taken me some time to do so. @kratman is attempting the unvendoring of the solver in https://github.com/pybamm-team/pybammsolvers/pull/4/, it might be good to migrate this PR there once that is done, @cringeyburger? It would make things a bit easier that were difficult to do here, especially with the troubles that were noticed with the benchmarks and the WSL-specific Dockerfile errors.

agriyakhetarpal commented:
> Yes, it would be great if we unvendor the solver and then work on migration.
> 
> I just wanted to point out that the benchmarks and dockerfile are fixed, along with everything else, atleast before I started the streak of merging commits from main. With this unvendoring, it would make the migration neat and clean with the new changes from main.

Thanks! Yes, sorry, I, too, meant that the migration would be cleaner with it, not that they were not working right now. Let's wait until that gets merged and we can alter your final report if you'd like to. Do you know why the `CIBW_BEFORE_ALL` step is failing? I'm unable to find out from the logs, but I feel it's getting stuck with downloading something. The Windows wheel failure looks fixable, though.

cringeyburger commented:
I tried fixing things, here is what I found:
1. After we incorporated `uv` in the workflow, I noticed that the logs for MacOS builds are too huge, and no matter how or what I do (even tried downloading them), I can't get to the bottom part. 
2. Even though `wheel.cmake` is set to `true` by default and we are explicitly setting it too, `scikit-build-core` refuses to run CMake before making the wheel. I tried using the override method - set `READTHEDOCS` env to `true`, and set `wheel.cmake` to true in the overrides, and it builds a proper wheel after running CMake.

agriyakhetarpal commented:
Thanks for taking a look – I think we should pin `scikit-build-core` to the last working version (maybe 0.9.10). Breakages like these will be troublesome for us to fix, and it might be better to always keep it pinned until it becomes a stable enough backend.

As for the macOS wheel builds, I got to notice them in live-action mode just now and I think there is an endless loop somewhere, since the before-all step kicks off another before-all step, and the next one, and the next, and so on until we run out of memory designated by the GHA runner. It's unlikely this is a bug on `cibuildwheel`'s side, though.

agriyakhetarpal commented:
It's because of this line: `python -m cibuildwheel --output-dir wheelhouse`, which triggers another `cibuildwheel` instance inside `cibuildwheel` itself. I guess you brought it back inadvertently at some place when merging changes from `develop`.

kratman commented:
@cringeyburger I think due to the separation of IDAKLU in #4487, we should probably move this work to https://github.com/pybamm-team/pybammsolvers

This will make the changes much simpler and easier to manage

kratman commented:
Changes migrated to pybammsolvers, closing for now
